File: python/백준/그래프/다이젝스트라/10282 해킹.py
import sys
import heapq

input = sys.stdin.readline
INF = int(1e9)

T = int(input())

for _ in range(T):
    n, d, c = map(int, input().split())
    
    graph = [[] for _ in range(n + 1)]
    distance = [INF] * (n + 1)
    
    for _ in range(d):
        a, b, s = map(int, input().split())
        graph[b].append((a, s))
    
    def dijkstra(start):
        queue = []
        heapq.heappush(queue, (0, start))
        distance[start] = 0
        
        while queue:
            dist, now = heapq.heappop(queue)
            
            if dist > distance[now]:
                continue
            
            for next_node, cost in graph[now]:
                new_cost = dist + cost
                
                if new_cost < distance[next_node]:
                    distance[next_node] = new_cost
                    heapq.heappush(queue, (new_cost, next_node))
    
    dijkstra(c)
    
    # 🔥 결과 계산
    count = 0
    max_time = 0
    
    for i in range(1, n + 1):
        if distance[i] != INF:
            count += 1
            max_time = max(max_time, distance[i])
    
    print(count, max_time)

File: cpp/백준/그래프/다이젝스트라/10282 해킹.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>
#include <algorithm>
using namespace std;

const int INF = 1e9;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n, d, c;
        cin >> n >> d >> c;

        vector<vector<pair<int, int>>> graph(n + 1);
        vector<int> distance(n + 1, INF);

        for (int i = 0; i < d; ++i) {
            int a, b, s;
            cin >> a >> b >> s;
            graph[b].push_back({a, s});
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> queue;
        queue.push({0, c});
        distance[c] = 0;

        while (!queue.empty()) {
            auto [dist, now] = queue.top();
            queue.pop();

            if (dist > distance[now]) continue;

            for (auto [next_node, cost] : graph[now]) {
                int new_cost = dist + cost;
                if (new_cost < distance[next_node]) {
                    distance[next_node] = new_cost;
                    queue.push({new_cost, next_node});
                }
            }
        }

        int count = 0;
        int max_time = 0;
        for (int i = 1; i <= n; ++i) {
            if (distance[i] != INF) {
                count += 1;
                max_time = max(max_time, distance[i]);
            }
        }

        cout << count << " " << max_time << "\n";
    }
    return 0;
}
